src/CodeGen_Hexagon.cpp

```diff
@@ -2734,16 +2739,23 @@ void CodeGen_Hexagon::visit(const Load *op) {
             CodeGen_Posix::visit(op);
             return;
           }
-          int offset = b->value;
-          if (mod_rem.remainder != ((bytes_in_vector() + offset) % bytes_in_vector())) {
+          int offset = (b->value % lanes);
+          // If the offset is negative, it is because we are trying to load something like
+          // f(x-1). In this case, the remainder should be (vector_width + offset) % vector_width
+          // Or in general for any type, it would be (lanes + offset) % lanes.
+          if (mod_rem.remainder != (lanes + offset) % lanes) {
             CodeGen_Posix::visit(op);
             return;
           }
-          int bytes_off = std::abs(offset) * (op->type.bits() / 8);
-          Expr base_low = offset > 0 ? add->a : simplify(add->a - width);
-          Expr base_high = offset > 0 ? simplify(add->a + width) : add->a;
-          Expr ramp_low = Ramp::make(base_low, 1, width);
-          Expr ramp_high = Ramp::make(base_high, 1, width);
+          // offset tells us that we are off by those many elements from the vector
+          // width. For e.g. a value of -1 when we are loading a half-word vector
+          // means we are reading one half-word left of an aligned address.
+          // so we are 1 * (64 / 32) = 2 bytes off (assuming 64 byte mode).
+          int bytes_off = std::abs(offset) * (native_vector_bytes / lanes);
```

Comment:
If it would make those indistinguishable, then how do you currently distinguish between ```ramp(a*64 + 1, 1, 64)``` and ```ramp(a*64 + 65, 1, 64)```? Looks like you need to also divide b->value by lanes and add that to base (use div_imp to have the right behavior relative to mod_imp).

As far as I can tell, for this purpose ```%``` and ```/``` are just slightly broken versions of ```mod_imp``` and ```div_imp```.